ibis/pandas/tests/test_udf.py

```diff
@@ -0,0 +1,77 @@
+import numpy as np
+
+import pandas as pd
+import pandas.util.testing as tm
+
+import ibis
+import ibis.expr.types as ir
+import ibis.expr.datatypes as dt
+
+from ibis.pandas.udf import udf, udaf
+from ibis.pandas.dispatch import pause_ordering
+
+
+with pause_ordering():
+
+    @udf(input_type=[dt.string], output_type=dt.int64)
+    def my_string_length(series, **kwargs):
+        return series.str.len() * 2
+
+    @udaf(input_type=[dt.string], output_type=dt.int64)
+    def my_string_length_sum(series, **kwargs):
+        return (series.str.len() * 2).sum()
+
+    @udaf(input_type=[dt.double, dt.double], output_type=dt.double)
+    def my_corr(lhs, rhs, **kwargs):
+        return lhs.corr(rhs)
+
+
+def test_udf():
+    df = pd.DataFrame({'a': list('abc')})
+    con = ibis.pandas.connect({'df': df})
+    t = con.table('df')
+    expr = my_string_length(t.a)
+
+    assert isinstance(expr, ir.Expr)
+
+    result = expr.execute()
+    expected = t.a.execute().str.len().mul(2)
+    tm.assert_series_equal(result, expected)
+
+
+def test_udaf():
+    df = pd.DataFrame({'a': list('cba')})
+    con = ibis.pandas.connect({'df': df})
+    t = con.table('df')
+    expr = my_string_length_sum(t.a)
+
+    assert isinstance(expr, ir.Expr)
+
+    result = expr.execute()
+    expected = t.a.execute().str.len().mul(2).sum()
+    assert result == expected
+
+
```

Comment:
could add some test for invalid udf decorator construction ? eg for passing invalid types
are there other ways these could fail?

Reply:
There are definitely not enough tests here, generally speaking.